chia/full_node/full_node.py

```diff
@@ -1121,69 +1136,119 @@ async def fetch_block_batches(
+            except Exception:
+                self.log.exception("Exception validating")
+            finally:
+                # finished signal with None
+                await output_queue.put(None)
+
+        async def ingest_blocks(
+            input_queue: asyncio.Queue[
+                Optional[
+                    tuple[WSChiaConnection, ValidationState, list[Awaitable[PreValidationResult]], list[FullBlock]]
+                ]
+            ],
         ) -> None:
             nonlocal fork_info
             block_rate = 0
             block_rate_time = time.monotonic()
             block_rate_height = -1
             while True:
-                res: Optional[tuple[WSChiaConnection, list[FullBlock]]] = await inner_batch_queue.get()
+                res = await input_queue.get()
                 if res is None:
-                    self.log.debug("done fetching blocks")
+                    self.log.debug("done validating blocks")
                     return None
-                peer, blocks = res
+                peer, vs, futures, blocks = res
                 start_height = blocks[0].height
                 end_height = blocks[-1].height
 
                 if block_rate_height == -1:
                     block_rate_height = start_height
 
-                # in case we're validating a reorg fork (i.e. not extending the
-                # main chain), we need to record the coin set from that fork in
-                # fork_info. Otherwise validation is very expensive, especially
-                # for deep reorgs
-                peak: Optional[BlockRecord]
-                if fork_info is None:
-                    peak = self.blockchain.get_peak()
-                    extending_main_chain: bool = peak is None or (
-                        peak.header_hash == blocks[0].prev_header_hash or peak.header_hash == blocks[0].header_hash
-                    )
-                    # if we're simply extending the main chain, it's important
-                    # *not* to pass in a ForkInfo object, as it can potentially
-                    # accrue a large state (with no value, since we can validate
-                    # against the CoinStore)
-                    if not extending_main_chain:
-                        if fork_point_height == 0:
-                            fork_info = ForkInfo(-1, -1, self.constants.GENESIS_CHALLENGE)
-                        else:
-                            fork_hash = self.blockchain.height_to_hash(uint32(fork_point_height - 1))
-                            assert fork_hash is not None
-                            fork_info = ForkInfo(fork_point_height - 1, fork_point_height - 1, fork_hash)
-
-                # The ValidationState object (vs) is an in-out parameter. the add_block_batch()
-                # call will update it
-                success, state_change_summary, err = await self.add_block_batch(
+                pre_validation_results = list(await asyncio.gather(*futures))
```

Comment:
something im missing here, what keeps the order of the blocks we are prevalidating ?

Reply:
we queue them in the correct order. It doesn't matter if they complete out-of-order, since we `await` the futures in-order

Reply:
or are you suggesting `gather()` can return the results in a different order than they're passed in?
This is how we gather these futures today (in `main`). The [documentation ](https://docs.python.org/3/library/asyncio-task.html#asyncio.gather)suggests that order is preserved.
>  The order of result values corresponds to the order of awaitables in aws.

Reply:
ok i thought gather returns on completion but i see now that we are waiting for all of them to complete, it ok if it keeps the order but i would add a comment since this can break the sync